ksqldb-engine/src/main/java/io/confluent/ksql/util/BinPackedStreamsMetadata.java

```diff
@@ -0,0 +1,185 @@
+    } finally {
+      // If error classification throws then we consider the error to be an UNKNOWN error.
+      // We notify listeners and add the error to the errors queue in the finally block to ensure
+      // all listeners and consumers of the error queue (e.g. the API) can see the error. Similarly,
+      // log in finally block to make sure that if there's ever an error in the classification
+      // we still get this in our logs.
+      final QueryError queryError =
+          new QueryError(
+              System.currentTimeMillis(),
+              Throwables.getStackTraceAsString(e),
+              errorType
+          );
+      for (QueryMetadata.Listener listener: listeners) {
+        listener.onError(queryError);
+      }
+      queryErrors.add(queryError);
+      LOG.error(
+          "Unhandled exception caught in streams thread {}. ({})",
+          Thread.currentThread().getName(),
+          errorType,
+          e
+               );
+    }
+    return StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse.REPLACE_THREAD;
+  }
+
+  public void addListener(QueryMetadata.Listener listener) {
+    listeners.add(listener);
+  }
+
+  public KafkaStreams getKafkaStreams() {
+    return kafkaStreams;
+  }
+
+  public KafkaStreams.State state() {
+    return kafkaStreams.state();
+  }
+
+  public Collection<StreamsMetadata> allMetadata() {
+    return kafkaStreams.allMetadata();
+  }
+
+  public Set<StreamsTaskMetadata> getTaskMetadata() {
+    return kafkaStreams.localThreadsMetadata()
+                       .stream()
+                       .flatMap(t -> t.activeTasks().stream())
+                       .map(StreamsTaskMetadata::fromStreamsTaskMetadata)
+                       .collect(Collectors.toSet());
+  }
+
+  public void addToplogy(final Topology topology) {
+    queryMap.put(topology.getName(), topology);
+    kafkaStreams.addTopology(topology);
+  }
+
+  public void stop(final QueryId queryId) {
+    kafkaStreams.removeTopology(queryId);
+  }
+
+  public void restart(final QueryId queryId) {
+    final KafkaStreams newKafkaStreams = buildKafkaStreams();
+    for (Topology topology: queryMap.values()) {
+      newKafkaStreams.addTopology(topology);
+    }
+    newKafkaStreams.start();
+    kafkaStreams.close();
+    kafkaStreams = newKafkaStreams;
+  }
+
+  public boolean isError(final QueryId queryId) {
+    return !queryErrors.toImmutableList().isEmpty();
+  }
+
+  public void close(final QueryId queryId) {
+    queryMap.remove(queryId.toString());
+    kafkaStreams.removeTopology(queryId);
+  }
+
+  public void start(final QueryId queryId) {
+    kafkaStreams.addTopology(queryMap.get(queryId.toString()));
+  }
+
+  public List<QueryError> getQueryErrors(final QueryId queryId) {
+    return queryErrors.toImmutableList();
+  }
+
+  public Map<String, Map<Integer, LagInfo>> allLocalStorePartitionLags(final QueryId queryId) {
+    try {
+      return kafkaStreams.allLocalStorePartitionLags();
```

Comment:
Maybe I can filter for the queries but am not sure

Reply:
Yeah I think we should filter for the queries --- from the queryId we get the named topology and all its contained taskIds, and filter based on the taskID -> partitions.

Reply:
We will have to add the tasks to the queryId I think but that might be something we do later